[BEAM-7760] Added Interactive Beam module

KevinGG opened the pull request:
**Please** add a meaningful description for your change here
1. Added interactive_beam module that will serve sugar syntax and
shorthand functions to apply interactivity and visualize 
PCollection data.
2. This commit implemented the implicitly managed Interactive Beam
environment to track definition of user pipelines. It exposed a watch()
interface for users to explicitly instruct Interactive Beam the 
whereabout of their pipeline definition when it's not in __main__.
3. This commit exposed a PCollection data exploration interface
visualize(). Implementation is yet to be added.
4. Added interactive_environment module for internal usage without
backward-compatibility. It holds the cache manager and watchable
metadata for current interactive environment/session/context. Interfaces
are provided to interact with the environment and its components.
5. Unit tests included.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

KevinGG commented:
R: @aaltay 
retest this please

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,199 @@
+  current_env().watch(watchable)
+
+
+def create_pipeline(runner=None, options=None, argv=None):
```
This is not really very different than beam.Pipeline(...). It might be better to avoid adding. Using existing constructs will have the advantage of existing documentation and user knowledge, and user's can copy contents out of notebooks without much change.

KevinGG replied:
I'll make the change to only use interactive runner (and by default the underlying runner with direct runner) for now.

The function basically creates a pipeline object that is interactive. Currently we could only provide interactive experience if the underlying runner is DirectRunner (which makes the code equivalent to `beam.Pipeline(interactive_runner.InteractiveRunner(), ...))`.
Yeah, you're right, this is confusing. I'll remove the runner arg from the signature. In the future, when more underlying runners are supported, we can add back the runner arg and set the runner in the InteractiveRunner()
as `return beam.Pipeline(interactive_runner.InteractiveRunner(underlying_runner=runner))`

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,199 @@
+
+def visualize(pcoll):
+  """Visualizes a PCollection."""
+  # TODO(ningk)
```
Perhaps, start a new tracking top level JIRA and starting adding blocking JIRAs for these todos.

KevinGG replied:
Will do that and replace `TODO(ningk)` with `TODO(BEAM-XXXX)`

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,199 @@
+  pass
+
+
+def run_pipeline(pipeline, runner=None, options=None):
```
What would this do? Since create_pipeline above will set the runner and options already for the pipeline.

KevinGG replied:
The run_pipeline takes in an existing interactive pipeline object, a new runner and options for that new runner so that user doesn't need to change/modify/re-execute any code in notebook to run the pipeline on selected platform say Dataflow by adding a new notebook cell and execute the single line of code.

We'll just have a variation of run_with_dataflow() at this time. I can replace this with run_with_dataflow() if it resolves confusion.

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,199 @@
+  return current_env(cache_manager)
+
+
+class InteractiveEnv(object):
```
Should we spell out InteractiveEnv to InteractiveEnvironment?

KevinGG replied:
Yes, we can. And I really want to hide this class from the interactive Beam (iBeam) user. I'll move everything about the environment to a different module.
This is supposed to be used by developers of iBeam internally.
iBeam users shouldn't use or care about what the implementation is.

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,199 @@
+    else:
+      self._watching_set.add(watchable)
+
+  def watching(self):
```
Could you add pydocs to all public methods?

KevinGG replied:
Will do.

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,199 @@
+
+
+def watch(watchable):
+  """Watches a watchable so that iBeam can understand your pipeline.
```
Pydocs have quite a bit of documentation. I suggest reviewing with @rosetn for consistency.

KevinGG commented:
Updated the PR with 2nd commit, also sent out an email briefing the interactive Beam work for this PR and future plan.

We should expect many PRs in the near future re-writing the interactive Beam (how cache used, how DOT renders, streaming has different interactive behavior from batch) since we changed how the underlying magic caching works.

The other main work is PCollection Data visualization.

However, except the interactive_beam module, everything else should not be concerned by either Beam users or developers. The change is within interactive Beam scope and any magic implemented is implicit. We shall update the README under interactive package as work evolves. 

PTAL
R: @aaltay

aaltay commented:
As an update, I would like the dev@ discussion to reach to a consensus state fist. Adding a README sounds good. And you might want to share a broader design document with dev@ including plans for future potential changes.

KevinGG commented:
Cool! I'll not touch README yet since I'm constructing the building blocks of a new iBeam without integrating (thus changing) behaviors of existing iBeam. But once I make those integration and change, I'll update the README as changes go.

For a broader design document, I composed a globally visible [design](https://docs.google.com/document/d/1DYWrT6GL_qDCXhRMoxpjinlVAfHeVilK5Mtf8gO6zxQ/edit?usp=sharing) overview describing changes we are making to components around interactive runner. I'll share the document in our email thread too.

The truth is since interactive runner is not yet a recognized runner as part of the Beam SDK (and it's fundamentally a wrapper around direct runner), we are not touching any Beam SDK components. We'll not change any behavior of existing Beam SDK and we'll try our best to keep it that way in the future.

In the mean time, I'll work on other components orthogonal to Beam such as Pipeline Display and Data Visualization I mentioned in the design overview.

Thanks!

KevinGG commented:
Hi all, I'll leave another 3 days for [design](https://docs.google.com/document/d/1DYWrT6GL_qDCXhRMoxpjinlVAfHeVilK5Mtf8gO6zxQ/edit?usp=sharing) review. Then we can have a vote session if there is no objections.

Thanks!

KevinGG commented:
Based on our offline discussion, we've decided not to make create_pipeline() and run_pipeline() API as part of interactive_beam module, but incorporating the features into existing beam module. I've updated the design doc: https://docs.google.com/document/d/1DYWrT6GL_qDCXhRMoxpjinlVAfHeVilK5Mtf8gO6zxQ/edit?usp=sharing

I've also removed the related code and tests from this PR.

In future beam changes, 
1. the beam.Pipeline() constructor would by default use runner=InteractiveRunner() when `apache_beam.runners.interactive.interactive_beam` module has been imported.
2. the pipeline.run() function would be able to take in runner=? and options=? to run existing pipeline with selected runner and options. 'runner' option would support both string name and runner object.

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+def watch(watchable):
+  """Watches a watchable so that Interactive Beam understands your pipeline.
+
+  If you write Beam pipeline in a notebook or __main__ module directly, since
```
For a follow up PR, I will suggest working with @rosetn to improve this comment. This comment will serve as documentation, and it will help to get it reviewed by tech writers.

KevinGG replied:
Sure, I'll contact @rosetn directly to work on improve the documentation. Would you please share me their ldap or work email address so that I can reach out?

KevinGG replied:
Thx for sharing the contact!
Will wait for @rosetn 's feedback to improve the comments before merging. ETA by this Friday.

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_beam_test.py:
```diff
@@ -0,0 +1,70 @@
+    test_env = ie.InteractiveEnvironment()
+    ib.watch(_module_name)
+    test_env.watch(_module_name)
+    self.assertEqual(ie.current_env().watching(), test_env.watching())
```
What is really tested here? We call watch on two environments and check that they have the same results. It is not checking whether the _module_name is in the watch set or not as the test name implies. (Same comment for other assertEqual's in this file.)

KevinGG replied:
These tests verifies that the static watch() function in interactive_beam module will eventually generate the same watching() items to the member watch() function in interactive_environment module InteractiveEnvironment class when the correct InteractiveEnvironment object is accessed.
Whatever future changes happen to the wrapper in interactiv_beam module, it should keep this behavior.
It currently basically tests if "ie.current_env().watch(watchable)" works as intended: create environment if absent, access current environment and invoke the watch() logic for that environment. 
It doesn't verify if "watch(watchable)" generates expected "watching()" because that is not the logic of this module.

The unit tests test the wrapped watch() logic in interactive_beam module where the InteractiveEnvironment objects may be created and accessed. The real watch() logic within an InteractiveEnvironment object from interactive_environment module is tested in its own unit test.

aaltay replied:
Sounds good.

aaltay commented on sdks/python/apache_beam/runners/interactive/interactive_environment.py:
```diff
@@ -0,0 +1,106 @@
+"""Module of the current Interactive Beam environment.
+
+Provides interfaces to interact with existing Interactive Beam environment.
+Internally used by Interactive Beam. External Interactive Beam users please use
```
For marking internal use, we generally use "For internal use only; no backwards-compatibility guarantees." on its own line.

See: https://github.com/apache/beam/blob/51a13f0a8e5a1088f8a96942b4787d7da25e54dc/sdks/python/apache_beam/utils/counters.py#L23

KevinGG replied:
Applied the convention.

aaltay commented:
Could you check the failing test please?

rosetn commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+
+    Interactive Beam will cache init_pcoll for the first run. You can use:
+
+    visualize(init_pcoll)
```
It might be easier to see related ideas if you add some transitions:

"Interactive Beam caches init_pcoll for the first run. Once the pipeline is executed, you can use visualize(init_pcoll) to visualize data from init_pcoll. 

You can then make the following change to your original pipeline to add squares using the init_pcoll data:
 squares = init_pcoll | 'Square' >> beam.Map(lambda x: x*x)
"

I'm not sure why adding squares will add an "interactive experience" though, could you elaborate a bit more?

KevinGG replied:
I'll remove this part. It's not relevant to `watch()`.  The interactivity happens when the user incrementally adding cells to develop further pipeline from existing pipeline object without re-executing previous cells. And optimization to pipeline run is done so that previous parts of the pipeline are not re-executed.

rosetn commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+    When you re-run the pipeline from the line you just added, squares will
+    use the init_pcoll data cached so you can have an interactive experience.
+
+  Currently the implementation mainly watches for PCollection variables defined
```
I think this information needs to happen first--what do you think about moving lines 68-80 to line 38? If so, you can disregard my comments above about "understands"

KevinGG replied:
Yes! Moving it up.

KevinGG commented:
> Could you check the failing test please?

Thanks! Sure.
Those are build failures from 3 tasks:
- :sdks:python:test-suites:tox:py2:docs
- :sdks:python:test-suites:tox:py2:lintPy27
- :sdks:python:test-suites:tox:py2:lintPy27_3
I'll update the docstring and comments following rosetn@ 's suggestions and format everything to pass these tasks.

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+about how hidden states in the interactive session are managed.
+
+Note: Backward-compatibility of Interactive Beam is only guaranteed within this
+module. Please only invoke interfaces provided by this module in your notebook
```
Thanks! I'll apply
> "Note: If you want backward-compatibility, only invoke interfaces provided by this module in your notebook or application code."

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+
+
+def watch(watchable):
+  """Watches a watchable so that Interactive Beam understands your pipeline.
```
Sure!

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+
+
+def watch(watchable):
+  """Watches a watchable so that Interactive Beam understands your pipeline.
```
I'll move the later explanation of watchable to the front just beneath this line.

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+
+
+def watch(watchable):
+  """Watches a watchable so that Interactive Beam understands your pipeline.
```
Yes, it means "so that Interactive Beam has information on where your pipelines are defined".

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+def watch(watchable):
+  """Watches a watchable so that Interactive Beam understands your pipeline.
+
+  If you write Beam pipeline in a notebook or __main__ module directly, since
```
Applied.

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+  Interactive Beam. However, if your Beam pipeline is defined in some module
+  other than __main__, e.g., inside a class function or a unit test, you can
+  watch() the scope to instruct the whereabouts of your pipeline definition so
+  Interactive Beam could apply interactivity to your pipeline when running it.
```
Added "that".

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+  """Watches a watchable so that Interactive Beam understands your pipeline.
+
+  If you write Beam pipeline in a notebook or __main__ module directly, since
+  __main__ module is always watched by default, you don't have to instruct
```
Done.

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+        return p
+    Foo().build_pipeline().run()
+
+    Interactive Beam will cache init_pcoll for the first run. You can use:
```
Changed. And I will apply it in the future. Thanks!

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+def watch(watchable):
+  """Watches a watchable so that Interactive Beam understands your pipeline.
+
+  If you write Beam pipeline in a notebook or __main__ module directly, since
```
I'll remove "a notebook or " to avoid the confusion. It doesn't matter what notebook the user uses. I think Python users know where __main__ module is even if they are using some notebook product such as a Jupyter notebook.

KevinGG commented on sdks/python/apache_beam/runners/interactive/interactive_beam.py:
```diff
@@ -0,0 +1,88 @@
+  If you write Beam pipeline in a notebook or __main__ module directly, since
+  __main__ module is always watched by default, you don't have to instruct
+  Interactive Beam. However, if your Beam pipeline is defined in some module
+  other than __main__, e.g., inside a class function or a unit test, you can
```
Applied.

KevinGG commented:
Resolved the failed checks.

- :sdks:python:test-suites:tox:py2:docs

Example code block in Python docstring needs to conform to some special format to be parsable to generate documentation. Applied [Google docstring Style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).


- :sdks:python:test-suites:tox:py2:lintPy27
- :sdks:python:test-suites:tox:py2:lintPy27_3

watch() tests without directly using local variables. Silenced the pylint warnings. Unused variables are intended in the tests.

KevinGG commented:
Rebased against upstream. The doc and check fixes are squashed into a 2nd commit. Will squash all when ready to merge.

aaltay commented:
Can you check the test failures?

KevinGG commented:
> Can you check the test failures?

Yes, applied a blank new line in-between differently styled imports to pass the lint check.
Squashed all and rebased against upstream.

KevinGG commented:
R:@aaltay
Rebased on upstream and passed all checks.
Could we merge it now? Thanks!
